sphinx_gallery/binder.py

```diff
@@ -95,17 +117,76 @@ def gen_binder_rst(fname, binder_conf):
     return rst
 
 
-def copy_binder_reqs(app):
+def copy_binder_files(app, exception):
+    """Copy all Binder requirements and notebooks files."""
+    if exception is not None:
+        return
+
+    if app.builder.name not in ['html', 'readthedocs']:
+        return
+
+    gallery_conf = app.config.sphinx_gallery_conf
+    binder_conf = check_binder_conf(gallery_conf.get('binder'))
+
+    # Copy the requirements files for binder
+    if not len(binder_conf) > 0:
+        return
+
+    logger.info('copying binder requirements...', color='white')
+    _copy_binder_reqs(app, binder_conf)
+    _copy_binder_notebooks(app)
+
+
+def _copy_binder_reqs(app, binder_conf):
     """Copy Binder requirements files to a "binder" folder in the docs."""
-    binder_conf = app.config.sphinx_gallery_conf['binder']
     path_reqs = binder_conf.get('dependencies')
 
-    binder_folder = os.path.join(app.builder.outdir, 'binder')
+    binder_folder = os.path.join(app.outdir, 'binder')
     if not os.path.isdir(binder_folder):
         os.makedirs(binder_folder)
     for path in path_reqs:
-        sh.copy(os.path.join(app.builder.srcdir, path),
-                binder_folder)
+        shutil.copy(os.path.join(app.srcdir, path), binder_folder)
+
+
+def _remove_ipynb_files(path, contents):
+    """Given a list of files in `contents`, remove all files named `ipynb` or
+    directories named `images` and return the result.
+
+    Used with the `shutil` "ignore" keyword to filter out non-ipynb files."""
+    contents_return = []
+    for entry in contents:
+        if entry.endswith('.ipynb'):
+            # Don't include ipynb files
+            pass
+        elif (entry != "images") and os.path.isdir(os.path.join(path, entry)):
+            # Don't include folders not called "images"
+            pass
+        else:
+            # Keep everything else
+            contents_return.append(entry)
+    return contents_return
+
+
+def _copy_binder_notebooks(app):
+    """Copy Jupyter notebooks to the binder notebooks directory.
+
+    Copy each output gallery directory structure but only including the
+    Jupyter notebook files."""
+
+    gallery_conf = app.config.sphinx_gallery_conf
+    gallery_dirs = gallery_conf.get('gallery_dirs')
+    binder_conf = gallery_conf.get('binder')
+    notebooks_dir = os.path.join(app.outdir, binder_conf.get('notebooks_dir'))
+    shutil.rmtree(notebooks_dir, True)
+    os.makedirs(notebooks_dir)
+
+    iterator = sphinx_compatibility.status_iterator(
+        gallery_dirs, 'copying binder notebooks...', length=len(gallery_dirs))
+
+    for i_folder in iterator:
+        shutil.copytree(os.path.join(app.srcdir, i_folder),
```

Comment:
In terms of readability I feel it would be clearer to walk the folder and copy all the .ipynb files explicitly. You would have to change the the beginning of the path from `app.srcdir` to `i_folder` by hand.

In other words if we only want to select `*.ipynb` files, I find the `ignore` API is not that well suited.

Reflecting about this though, I am wondering whether notebooks need other files to be fully functional. One example that comes to mind is if you do an import of a local python module (I seem to remember someone, maybe matplotlib, was doing it).

Reply:
I first proposed to use os.walk but after looking at @choldgraf's use of it I though it was a very bad idea to push forward, I needed few nested loops and you are joining paths all the time manually. I find the copytree implementation way cleaner. It copies the entire tree structure of the examples but only keeping the notebook files. It may feel counter intuitive to think about the ignore list but it's simpler to extend later on.
The ignore list is easier to extend and test as it is a separate function.

Regarding your use case of needing to copy other files that are not an example, the ignore list makes it easier. Unfortunately we will meet other roadblocks on that area regarding the way Sphinx-Gallery works. Examples are run in the `examples_dirs`, and thus look for file local to that place, in that place we put local modules, data files(like nilearn), or what ever else needed to run the script. But the `gallery_dirs` was only planned for web display, thus has only source code of executed file, rst representation, jupyter notebook representation and output images.

That means if we need to copy other files to give a full reproducible experience where we need local files, we would need to add an extra copy procedure to copy from the `examples_dirs`

Reply:
No offence is meant to @choldgraf for his implementation. That was probably the way to use os.walk, it just didn't look good to have it.

Reply:
No worries - I just want a clear and short path forward to get this PR merged. What shall we agree on for this particular point? Keep it the same?

Reply:
I think it is fine to keep it like this for now.